lib/TPP/GPU/LinalgToGpu.cpp

```diff
@@ -69,137 +71,213 @@ createGpuBlocksWrapper(Operation *op, ArrayRef<int64_t> blockDims,
+static int getComputeCapability(llvm::StringRef chip) {
+  llvm::StringRef delim = "_";
+  llvm::StringRef ccToken =
+      chip.substr(chip.find(delim) + delim.size(), chip.size());
+
+  return std::stoi(ccToken.str());
+}
+
+// Return true if hardware supports WMMA operations.
+static bool isMMASupported(LinalgToGpuOptions options) {
+  return (options.gpuTriple == "nvptx64-nvidia-cuda") &&
+         (getComputeCapability(options.gpuChip) >= 70);
+}
+
+// Helper struct containing hardware WMMA settings
+// such as tile sizes.
+struct WMMASettings {
+  int m;
+  int n;
+  int k;
+};
+
 // Return true if the operation can be represented with WMMA compute.
-static bool supportsMMACompute(linalg::LinalgOp linalgOp) {
-  if (!(isa_and_nonnull<linalg::MatmulOp>(linalgOp) ||
-        isa_and_nonnull<linalg::BatchReduceMatmulOp>(linalgOp))) {
-    return false;
+static std::optional<WMMASettings> getWMMASettings(linalg::LinalgOp linalgOp,
+                                                   LinalgToGpuOptions options,
+                                                   int kTile) {
+  if (!(isa<linalg::MatmulOp>(linalgOp) ||
+        isa<linalg::BatchReduceMatmulOp>(linalgOp))) {
+    return std::nullopt;
   }
 
   // Only static shapes are supported.
   if (linalgOp.hasDynamicShape())
-    return false;
+    return std::nullopt;
 
   auto aType = linalgOp.getDpsInputs()[0].getType().cast<ShapedType>();
   auto bType = linalgOp.getDpsInputs()[1].getType().cast<ShapedType>();
   auto cType = linalgOp.getDpsInits()[0].getType().cast<ShapedType>();
 
-  ArrayRef<int64_t> shapeA = aType.getShape();
-  ArrayRef<int64_t> shapeC = cType.getShape();
-  int64_t m = shapeC[0];
-  int64_t n = shapeC[1];
-  // Buffer A might be 2D (gemm) or 3D (brgemm) but the last dimension will
-  // always be reduction.
-  int64_t k = shapeA.back();
-
-  // For now, only M-N-K F16[16] x F16[16] x F16[16] WMMA variant is supported.
-  // TODO: add more WMMA combinations.
-  return aType.getElementType().isF16() && bType.getElementType().isF16() &&
-         cType.getElementType().isF16() && m == 16 && n == 16 && k == 16;
+  auto elemTypeA = aType.getElementType();
+  auto elemTypeB = bType.getElementType();
+  auto elemTypeC = cType.getElementType();
+
+  // TODO: Add more WMMA combinations.
+  bool isSupportedPrecision =
+      (elemTypeA.isF16() && elemTypeB.isF16() && elemTypeC.isF16()) ||
+      (elemTypeA.isF16() && elemTypeB.isF16() && elemTypeC.isF32());
+  if (!isSupportedPrecision)
+    return std::nullopt;
+
+  auto mDim = cType.getShape()[0];
+  auto nDim = cType.getShape()[1];
+  auto kDim = aType.getShape().back();
+
+  // Choose WMMA tile sizes.
+  // The computation dimensions must fit into the tiles.
+  // Reduction dimension tile size has to be compatible
+  // with the hardware sizes.
+  //
+  // TODO: Add more possible tile sizes and choose optimal one.
+  int wmmaTileM = 16;
+  int wmmaTileN = 16;
+  int wmmaTileK = 16;
```

Comment:
TODO: move this to a pass option provided by a user

It will be more flexible and provide space for easier experimentation in further lowering.